Add website (#42)
neuroinformatics-unit/derotation: docs/source/user_guide/key_concepts.md

Reviewer: This key concepts guide is good, but to aid the reader, I would split this up so that this page outlines the concepts themselves, and the code examples are elsewhere.

(user_guide/key_concepts)=
# Key Concepts

`derotation` is a modular package for reconstructing rotating image stacks acquired via line scanning microscopes. This page covers the core ideas, flow of information, and how the main modules interact.
See the [API documentation](../api_index.rst) for a detailed overview of the package.

## How derotation works

Without derotation, movies acquired with a line scanning microscope are geometrically distorted — each line is captured at a slightly different angle, making it hard to register or interpret the resulting images.

```{figure} ../_static/derotation_index.png

> In this figure you can see on the left a schematic of a line-scanning microscope acquiring an image of a grid. The scanning pattern plus the sample rotation lead fan-like aftefacts. On the right you see a grid that has been imaged while still (top), while rotating (middle) and the derotated image (bottom). The grid is now perfectly aligned.

Reviewer: ```suggestion
**Left:** schematic of a line-scanning microscope acquiring an image of a grid. The scanning pattern plus the sample rotation lead fan-like aftefacts. **Right:** grid that has been imaged while still (top), while rotating (middle) and the derotated image (bottom). The grid is now perfectly aligned.
```

```

If the angle of rotation is recorded, `derotation` can reconstruct each frame by assigning a rotation angle to each acquired line and rotating it back to its original position. This is what is called derotation-by-line. This process incrementally reconstructs each frame and can optionally include shear deformation correction.

```{figure} ../_static/derotation_by_line.gif

> In this animation you can see the reconstruction of a frame with derotation. The original frame with deformation is on the left. With subsequent iterations, derotation picks a line, rotates it according to the calculated angle, and adds it to the new derotated frame. The final result is on the right.

Reviewer: ```suggestion
Reconstruction of a frame with derotation. The original frame with deformation is on the left. With subsequent iterations, derotation picks a line, rotates it according to the calculated angle, and adds it to the new derotated frame. The final result is on the right.
```

### 2. **Full and Incremental Pipeline Classes**

Derotation provides two pre-built pipelines for end-to-end processing:

- **{class}`derotation.analysis.full_derotation_pipeline.FullPipeline`**
  Assumes randomized, alternating clockwise and counter-clockwise rotations. It performs:
  - Analog signal parsing
  - Angle interpolation
  - Bayesian optimization for center estimation
  - Line-by-line derotation using {func}`derotation.derotate_by_line.derotate_an_image_array_line_by_line`

- **{class}`derotation.analysis.incremental_derotation_pipeline.IncrementalPipeline`**
  Assumes a continuous rotation performed in small increments. Inherits from `FullPipeline` but skips Bayesian optimization.
  Useful when the center of rotation is known or estimated differently.

Both pipelines accept a **configuration dictionary** (see the [configuration guide](./configuration.md)) and produce:
- A derotated TIFF stack
- A CSV file with rotation angles and metadata
- Debugging plots
- A text file containing the estimated optimal center of rotation
- Log files

You can create custom pipelines by subclassing `FullPipeline` and overriding the relevant methods.

> See the [usage example](../examples/pipeline_with_real_data.zip) for how to instantiate `FullPipeline`, run it, and inspect its attributes and outputs.

Reviewer: A lot of people won't be happy to download a random zip archive. Also, I would host this data on GIN rather than in the git repo.

Author: That's actually a typo, there's no zip file!

```{figure} ../_static/out_of_plane.png

> In this schematic you can see on the left a representation of out of imaging plane rotation paired with the position of a cell at 0deg, 90 deg and 180 deg. The right and center schematics offer a side and top view of the planes and their relationship with the ellipse that can be fitted to the cell positions.

Reviewer: ```suggestion
**Left:** a representation of out of imaging plane rotation paired with the position of a cell at 0deg, 90 deg and 180 deg. **Centre:** Side view and **Right:** top view top view of the planes and their relationship with the ellipse that can be fitted to the cell positions.
```

(user_guide/key_concepts)=
# Key Concepts

`derotation` is a modular package for reconstructing rotating image stacks acquired via line scanning microscopes. This page covers the core ideas, flow of information, and how the main modules interact.

> See the [API documentation](api_index) for full reference.

Author: This should be solved now that you can build the api index

> See the [API documentation](api_index) for full reference.

Reviewer: It didn't work for me, I think it should be something like `/api_index`

## How derotation works

Without derotation, movies acquired with a line scanning microscope are geometrically distorted — each line is captured at a slightly different angle, making it hard to register or interpret the resulting images.

If the angle of rotation is recorded, `derotation` can reconstruct each frame by assigning a rotation angle to each acquired line and rotating it back to its original position. This is what is called derotation-by-line. This process incrementally reconstructs each frame and can optionally include shear deformation correction.

> ![Derotation process](../_static/derotation_by_line.gif)

Author: I used the exact frame rates of the experiments! I kind of like it this "slow". I'll think about it

## How to use derotation

There are two main ways to use `derotation`:

### 1. **Low-level core function**

> Use `derotate_an_image_array_line_by_line` to derotate an image stack, given:

Reviewer: Can this link to the API docs?

Author: Good point, it should. Have to investigate how.

> - The original calcium imaging movie (expects only one imaging plane)

Reviewer: This isn't necessarily limited to Ca imaging right?

Author: True 🤔

- Rotation angle per line

This is ideal for testing and debugging with synthetic or preprocessed data.

### 2. **Full or incremental pipeline classes**
Use the pre-made pipelines to run end-to-end processing:

> - `FullPipeline`: assumes randomised complete clockwise and counter-clockwise rotations. It includes:

Reviewer: Can this link to the API docs?

Author: As above

  - Analog signal parsing
  - Angle interpolation
  - Bayesian optimization for center estimation
  - Derotation by line (calling `derotate_an_image_array_line_by_line`)

- `IncrementalPipeline`: assumes a continuous rotation performed in small increments. It inherits functionality from the `FullPipeline` but does not perform Bayesian optimization. It can be useful as an alternative way to estimate center of rotation.

Both pipelines accept a configuration dictionary (see the [configuration guide](configuration)) and output:
- Derotated TIFF
- CSV with rotation angles and metadata
- Debugging plots
- A text file with optimal center of rotation
- Logs

You can subclass `FullPipeline` to create custom pipelines by overwriting relevant methods.

> ### Example usage of FullPipeline:

Reviewer: Would this be better in the examples?

```python
from derotation.config.load_config import update_config_paths, load_config
from derotation.analysis.full_derotation_pipeline import FullPipeline

# Load the configuration file
pipeline_type = "full"
config = load_config(pipeline_type)
config = update_config_paths(
    config=config,
    tif_path="/my/data/image_stack.tif",
    bin_path="/my/data/signals.bin",
    dataset_path="/my/data/",
    kind=pipeline_type,
)

# Load data and run the pipeline
pipeline = FullPipeline(config)
pipeline()
```

---

## Finding the center of rotation

An accurate center of rotation is crucial for high-quality derotation. Even small errors can produce residual motion in the derotated movie — often visible as circles traced by stationary objects as cells.

> ![Residual motion](../_static/wrong_center.png)

Reviewer: Sphinx has some nice functionality for image captions which would look nice here.

Author: Have to look at them thaks!

In this picture, you can see in red the centers of one of the cells in the movie across different angles. Since the center is not correctly estimated, the cell appears to move in a circle.

Derotation offers two approaches for estimating the center:

> ### Bayesian optimization via FullPipeline

Reviewer: I would make these just bold rather than H3, it looks a bit weird following on from the sentence above.

> This method searches for the correct center of rotation by derotating the whole movie and minimizing a custom metric, computed through the function `ptd_of_most_detected_blob`. It requires the average image of the derotated movie at different rotations angles, and from them detects blobs, searches for the most frequent and calculates the Point-to-Point Distance (PTD) for it across blob centers at different rotation angles.

Reviewer: Again, can these functions link to the API docs?

It is robust but computationally expensive.

### Ellipse fitting via IncrementalPipeline

This method exploits the fact that incremental datasets rotate very slowly and smoothly. It works by:

- Detecting largest blob in the first frame
- Tracking its position across rotations
- Fitting an ellipse to the trajectory

The center of the ellipse is assumed to match the true center of rotation. This method fails when the cell stops being visible in certain rotation angles.

Once the center is estimated, it can be fed to the FullPipeline to derotate the whole movie.

> #### Example usage of IncrementalPipeline:

Reviewer: Again, examples?

Author: 😭
